File: Cabfareprediction.R
#Loading required libraries
library(lubridate)
library(ggplot2)
library(dplyr)
library(ggcorrplot)
library(corrplot)
library(scales)
library(psych)

#Gettings the files
train = read.csv("C:/Users/Ananya Verma/Documents/traincab.csv")
test = read.csv("C:/Users/Ananya Verma/Documents/testcab.csv")

#Making a copy of train variable for ease 
qwe = train


#Trying to understand the data better
dim(train)
dim(test)
head(train)
head(test)

#Trying to understand the fare_amount in train set
summary(train$fare_amount)

#The fare_amount variable has some wierd values like "(Other)"

#Finding the column names
colnames(train)
colnames(test)

##Dealing with pickup_datetime variable## 
#Seperating pickupdatetime variable into seperate variables
train = train %>% 
  mutate(
    pickup_datetime = ymd_hms(pickup_datetime),
    month = month(pickup_datetime),
    year = year(pickup_datetime),
    day = day(pickup_datetime),
    dayOfWeek = wday(pickup_datetime),
    hour = hour(pickup_datetime)
  )

#Similarly for test
test = test %>% 
  mutate(
    pickup_datetime = ymd_hms(pickup_datetime),
    month = month(pickup_datetime),
    year = year(pickup_datetime),
    day = day(pickup_datetime),
    dayOfWeek = wday(pickup_datetime),
    hour = hour(pickup_datetime)
  )

#Deleting pickup_datetime variable as it is of now use now
train$pickup_datetime = NULL

#Verifying the newly added columns
colnames(train)
colnames(test)

#Finding Missing values in data
missingval_train= data.frame(apply(train,2,function(x){sum(is.na(train))}))
missingval_train
#Since passenger_Count, month, year,day,hour,dayOfWeek has missing values, let us impute missing value with median
train$passenger_count[is.na(train$passenger_count)] = median(train$passenger_count, na.rm = T)
train$month[is.na(train$month)] = median(train$month, na.rm = T)
train$year[is.na(train$year)] = median(train$year, na.rm = T)
train$day[is.na(train$day)] = median(train$day, na.rm = T)
train$dayOfWeek[is.na(train$dayOfWeek)] = median(train$dayOfWeek, na.rm = T)
train$hour[is.na(train$hour)] = median(train$hour, na.rm = T)


#Checking if missing values in passenger count were treated right
missingval_train= data.frame(apply(train,2,function(x){sum(is.na(train))}))
missingval_train

View(train)
#0 missing values

#Moving forward with test data
missingval_test= data.frame(apply(test,2,function(x){sum(is.na(test))}))
missingval_test
#We see that test data has zero missing values

#Finding NA's in data
cat('NAs in train dataset:', sum(is.na(train)), '\nNAs in train dataset:', sum(is.na(train)), sep = ' ')
cat('NAs in test dataset:', sum(is.na(test)), '\nNAs in test dataset:', sum(is.na(test)), sep = ' ')

#Finding type of data
class(train)
class(test)

#The row 1124 has fare_amount "430-" hence we will try to remove it
train[1124,]
train = train[-c(1124), ]
##we have deleted "430-" value and then stored the new data in "train"

##Treating Longitudes and Latitudes##
#Trying to understand the latitudes and longitude values in train set
summary(train$pickup_latitude)
summary(train$dropoff_latitude)
del_Lat_max = which(train$pickup_latitude > 90 )
train = train[-del_Lat_max,]
del_Lat_min = which(train$pickup_latitude < -90)
#no values satisfying above condition

#Removing the rows where drop off latitude is > 90 and < -90.
del_Lat_max = which(train$dropoff_latitude > 90)
#no values satisfying above condition
del_Lat_min = which(train$dropoff_latitude < -90)
#no values satisfying above condition

# Getting the summary of pick up and Drop off latitude after clean up 
summary(train$pickup_latitude)
summary(train$dropoff_latitude)


#Checking if any longitude is < -180 or > 180
summary(train$pickup_longitude)
summary(train$dropoff_longitude)

#Removing the rows where pick up longitude is < -180 and > 180
pick_longi_min = which(train$pickup_longitude < -180)
#no values satisfying above condition
pick_longi_max = which(train$pickup_longitude > 180)
#no values satisfying above condition

#Removing the rows where drop off longitude is < -180 and > 180
drop_longi_min = which(train$dropoff_longitude < -180)
drop_longi_max = which(train$dropoff_longitude > 180)
#no values satisfying above conditions

# Getting the summary of pick up and Drop off longitude after clean up 
summary(train$pickup_longitude)
summary(train$dropoff_longitude)


#Checking train and test set with new variables
head(train)
head(test)

#The train dataset has negative values in fare_amount which is not practical, hence removing those values
train$fare_amount = as.numeric(train$fare_amount)
str(train$fare_amount)
negativefare = which(train$fare_amount < 0)
summary(train$fare_amount)
#Checking fare_amount for negative values, min value is 1 which means no negative value existing.
#train_data = train_data[-negativefare, ]

class(train) #data.frame

#--------------------VIsualization------------------------------------------------------
colnames(train)

ggplot(train, aes(x=day,y = fare_amount))+geom_bar(stat = "identity",na.rm = FALSE, fill="steelblue",show.legend = T, width = 0.9)+xlab("Day")+ylab("fare_amount")+scale_x_discrete(breaks=pretty_breaks(n=10), )+scale_y_discrete(breaks=pretty_breaks(n=10))+theme(text = element_text(size = 15))+geom_text(aes(label=train$day, vjust=-0.3))
ggplot(train, aes(x=dayOfWeek,y = fare_amount)) +geom_bar(stat = "identity",na.rm = FALSE, fill="steelblue",show.legend = T, width = 0.9)+xlab("Day Of week")+ylab("fare_amount")+scale_x_discrete(breaks=pretty_breaks(n=10), )+scale_y_discrete(breaks=pretty_breaks(n=10))+theme(text = element_text(size = 15))+geom_text(aes(label=train$dayOfWeek, vjust=-0.3))
ggplot(train, aes(x=hour,y = fare_amount)) +geom_bar(stat = "identity",na.rm = FALSE, fill="steelblue",show.legend = T, width = 0.9)+xlab("Hour")+ylab("fare_amount")+scale_x_discrete(breaks=pretty_breaks(n=10), )+scale_y_discrete(breaks=pretty_breaks(n=10))+theme(text = element_text(size = 15))+geom_text(aes(label=train$hour, vjust=-0.3))
ggplot(train, aes(x=year,y = fare_amount)) +geom_bar(stat = "identity",na.rm = FALSE, fill="steelblue",show.legend = T, width = 0.9)+xlab("Year")+ylab("fare_amount")+scale_x_discrete(breaks=pretty_breaks(n=10), )+scale_y_discrete(breaks=pretty_breaks(n=10))+theme(text = element_text(size = 15))+geom_text(aes(label=train$year, vjust=-0.3))
ggplot(train, aes(x=month,y = fare_amount)) +geom_bar(stat = "identity",na.rm = FALSE, fill="steelblue",show.legend = T, width = 0.9)+xlab("Month")+ylab("fare_amount")+scale_x_discrete(breaks=pretty_breaks(n=10), )+scale_y_discrete(breaks=pretty_breaks(n=10))+theme(text = element_text(size = 15))+geom_text(aes(label=train$month, vjust=-0.3))
#ggplot(train, aes(x=passenger_count,y = fare_amount)) +geom_bar(stat = "identity", fill="green")+scale_x_discrete(breaks=5)
ggplot(train, aes(x=passenger_count,y = fare_amount)) +geom_point(aes_string(color=train$passenger_count))+theme_bw()+ylab("fare_Amount")+xlab("Passenger Count")+theme(text=element_text(size=10))+scale_x_continuous(breaks=pretty_breaks(n=10))+scale_y_continuous(breaks=pretty_breaks(n=10))
ggplot(train, aes(x=dropoff_latitude,y = fare_amount)) +geom_point(aes_string(color=train$dropoff_latitude))+theme_bw()+ylab("fare_Amount")+xlab("Dropoff Latitude")+theme(text=element_text(size=10))+scale_x_continuous(breaks=pretty_breaks(n=10))+scale_y_continuous(breaks=pretty_breaks(n=10))
ggplot(train, aes(x=dropoff_longitude,y = fare_amount)) +geom_point(aes_string(color=train$dropoff_longitude))+theme_bw()+ylab("fare_Amount")+xlab("Dropoff Longitude")+theme(text=element_text(size=10))+scale_x_continuous(breaks=pretty_breaks(n=10))+scale_y_continuous(breaks=pretty_breaks(n=10))
ggplot(train, aes(x=pickup_latitude,y = fare_amount)) +geom_point(aes_string(color=train$pickup_latitude))+theme_bw()+ylab("fare_Amount")+xlab("Pickupp Latitude")+theme(text=element_text(size=10))+scale_x_continuous(breaks=pretty_breaks(n=10))+scale_y_continuous(breaks=pretty_breaks(n=10))
ggplot(train, aes(x=pickup_longitude,y = fare_amount)) +geom_point(aes_string(color=train$pickup_longitude))+theme_bw()+ylab("fare_Amount")+xlab("Pickup Longitude")+theme(text=element_text(size=10))+scale_x_continuous(breaks=pretty_breaks(n=10))+scale_y_continuous(breaks=pretty_breaks(n=10))



#Outlier analysis
cnames= colnames(train)
cnames

#Outliers Visualization
for (i in 1:length(cnames))
{
  assign(paste0("an",i), ggplot(aes_string(y = (cnames[i]), x = "fare_amount"), data = subset(train))+
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="fare_amount")+
           ggtitle(paste("Box plot of responded for",cnames[i])))
}

library(gridExtra)
grid.arrange(an3, an5,an4, an6, an7, an8, an9,an10, an11,an1,an2, ncol=4)


#Treating Outliers

val = train$pickup_longitude[train$pickup_longitude %in% boxplot.stats(train$pickup_longitude)$out]
length(val)
train = train[which(!train$pickup_longitude %in% val),]
View(train)

#@@@@@@@@@@

val1 = train$passenger_count[train$passenger_count %in% boxplot.stats(train$passenger_count)$out]
length(val1)
train = train[which(!train$passenger_count %in% val1),]
View(train)

#@@@@@@@@@@@@@@@@@@@@@@@@
val2 = train$dropoff_longitude[train$dropoff_longitude %in% boxplot.stats(train$dropoff_longitude)$out]
length(val2)
train = train[which(!train$dropoff_longitude %in% val2),]
View(train)
#@@@@@@@@@@@@@@@@@@@@@@@@@@
val3 = train$dropoff_latitude[train$dropoff_latitude %in% boxplot.stats(train$dropoff_latitude)$out]
length(val3)
train = train[which(!train$dropoff_latitude %in% val3),]
View(train)
#@@@@@@@@@@@@@@@@@@@@@@@@@@@@
val4 = train$pickup_latitude[train$pickup_latitude %in% boxplot.stats(train$pickup_latitude)$out]
length(val4)
train = train[which(!train$pickup_latitude %in% val4),]
View(train)

#We have treated the outliers, We shall now move forward with linear regression model


colnames(train)

#Visualizations
#plotting bar graph to observe variable day
ggplot(train, aes(x=day,y = fare_amount)) +geom_bar(stat = "identity", fill="navyblue")+scale_x_continuous(breaks = pretty_breaks(n=10))+scale_y_continuous(breaks = pretty_breaks(n=10))

#plotting bar graph to observe variable year
ggplot(train, aes(x=year,y = fare_amount)) +geom_bar(stat = "identity", fill="lightblue")+scale_x_continuous(breaks = pretty_breaks(n=10))+scale_y_continuous(breaks = pretty_breaks(n=10))

#plotting bar graph to observe month
ggplot(train, aes(x=month,y = fare_amount)) +geom_bar(stat = "identity", fill="green")+scale_x_continuous(breaks = pretty_breaks(n=10))+scale_y_continuous(breaks = pretty_breaks(n=10))

#plotting bar graph to observe hout
ggplot(train, aes(x=hour,y = fare_amount)) +geom_bar(stat = "identity", fill="skyblue")+scale_x_continuous(breaks = pretty_breaks(n=10))+scale_y_continuous(breaks = pretty_breaks(n=10))
#plotting bar graph to observe pickup_latitude
ggplot(train,aes(x=pickup_latitude))+
  geom_histogram(fill="black",alpha=0.5) +
  ggtitle("Histogram of pickup_latitude")

#plotting bar graph to observe pickup_longitude
ggplot(train,aes(x=pickup_longitude))+
  geom_histogram(fill="yellow",alpha=0.5) +
  ggtitle("Histogram of pickup_longitude")

#plotting bar graph to observe day

ggplot(train, aes(x=day,y = passenger_count)) +geom_bar(stat = "identity")

#plotting bar graph to observe dropoff_latitude
ggplot(train,aes(x=dropoff_latitude))+
  geom_histogram(fill="brown",alpha=0.5) +
  ggtitle("Histogram of dropoff_latitude")


#plotting bar graph to observe dayOfWeek
ggplot(train,aes(x=dayOfWeek))+
  geom_histogram(fill="orange",alpha=0.5) +
  ggtitle("Histogram of dayOfWeek")


#plotting bar graph to observe dropoff_longitude
ggplot(train,aes(x=dropoff_longitude))+
  geom_histogram(fill="green",alpha=0.5) +
  ggtitle("Histogram of dropoff_longitude")


######################  Decision Tree Model   #################

fit = rpart(fare_amount~ ., data = train, method = "anova")
summary(fit)

#Predict for new test cases
predictions_DT = predict(fit, test)
predictions_DT

y=train$fare_amount#Storing actual values in a object for ease

#CHeck Performance of model
#MAPE
#calculate MAPE
MAPE = function(y, yhat){
  mean(abs((y - yhat)/y))
}
MAPE(y, predictions_DT)
#Error rate : 3.752733
#Accuracy : 96.25



########Linear Regression Model#############
#Check multicolinearity

library(usdm)
library(rpart)

#CHecking Multicollinearity
library(corrplot)
qwe= as.numeric(qwe)

vif(qwe)
lm_model = lm(fare_amount ~., data = train)

#Summary of the model
summary(lm_model)

#Predict
predictions_LR = predict(lm_model, test)

#Calculate MAPE
MAPE(train$fare_amount, predictions_LR)
#Error rate : 3.926267
#Accuracy : 96.07

File: Cabfareprediction.py
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from math import sqrt
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn import tree
from collections import Counter
from sklearn.svm import SVR
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
import seaborn as sns
from sklearn.naive_bayes import GaussianNB
import statsmodels.api as sm
from sklearn.tree import DecisionTreeRegressor


#Calling Train and Test datasets
train = pd.read_csv("C:/Users/Ananya Verma/Documents/traincab.csv")
test = pd.read_csv("C:/Users/Ananya Verma/Documents/testcab.csv")

#Looking at first 5 entries of train and test dataset
print(train.head())
print(test.head())

print(train['fare_amount'].describe())

#checking datatypes in train dataset
print(train.dtypes)
print(test.dtypes)

#we will try to find the index of '430-' value in fare_amount and will drop the value
print(train.index[train.fare_amount == '430-'])

#converting fare_amount to float 
train['fare_amount'] = pd.to_numeric(train['fare_amount'],errors='coerce')
train['fare_amount'].dtype

#Checking for Multicollinearity
corr = train.corr()
print(corr)
cmap = sns.diverging_palette(220, 10, as_cmap=True)
sns.heatmap(corr, cmap=cmap, vmax=.3, center=0, square=True, linewidths=.5, cbar_kws={"shrink": .5})
plt.show()

#Removing the negative values from the fare_amount variable
print('Old count: %d' % len(train))
train = train[train.fare_amount>=0]
print('New count: %d' % len(train))


#Checking the missing values in train and test dataset
missvaltrain = pd.DataFrame(train.isnull().sum())
missvaltest = pd.DataFrame(test.isnull().sum())

#Imputing missing values in passenger_count variable with median
train['passenger_count'] = train['passenger_count'].fillna(train["passenger_count"].median())

#Verifying the passenger count variable
print(train['passenger_count'])
print(pd.DataFrame(train.isnull().sum()))


#Split our Datetime into individual columns year, month, day,
#weekday, hour and then dropping datetime column
def aligndatetime(df):
    df["pickup_datetime"] = df["pickup_datetime"].map(lambda x: str(x)[:-3])
    df["pickup_datetime"] = pd.to_datetime(df["pickup_datetime"], format='%Y-%m-%d %H:%M:%S')
    df['year'] = df.pickup_datetime.dt.year
    df['month'] = df.pickup_datetime.dt.month
    df['day'] = df.pickup_datetime.dt.day
    df['weekday'] = df.pickup_datetime.dt.weekday
    df['hour'] = df.pickup_datetime.dt.hour
    return(df["pickup_datetime"].head())    
print(aligndatetime(train))
print(aligndatetime(test))

train.drop('pickup_datetime', axis=1, inplace=True)
test.drop('pickup_datetime', axis=1, inplace=True)

#Checking the datasets for new columns added in the last step
print(train.head())
print(test.head())

#Checking for NA in dataset
print(train.isnull().any())
print(test.isnull().any())
print(train.isnull().sum())
train=train.fillna(train.mean())
train.isnull().sum()

print(train.isnull().sum())

#Dealing with the logitude and latitude variables, we consider Manhattan distance
train['abs_long']=abs(train['pickup_longitude']-train['dropoff_longitude'])
train['abs_lat']=abs(train['pickup_latitude']-train['dropoff_latitude'])

#Checking the new variables abs_long and abs_lat
print(train.head())

#Performing the same activity for the test dataset
test['abs_long']=abs(test['pickup_longitude']-test['dropoff_longitude'])
test['abs_lat']=abs(test['pickup_latitude']-test['dropoff_latitude'])

#Checking the new variables abs_long and abs_lat
print(test.head())
print(train.columns)
print(test.columns)

train['fare_amount'] =  train['fare_amount'].astype(int)
train['pickup_longitude'] =  train['pickup_longitude'].astype(int)
train['pickup_latitude'] =  train['pickup_latitude'].astype(int)
train['dropoff_longitude'] =  train['dropoff_longitude'].astype(int)
train['dropoff_latitude'] =  train['dropoff_latitude'].astype(int)
train['year'] =  train['year'].astype(int)
train['passenger_count'] =  train['passenger_count'].astype(int)
train['month'] =  train['month'].astype(int)
train['day'] =  train['day'].astype(int)
train['hour'] =  train['hour'].astype(int)
train['weekday'] =  train['weekday'].astype(int)
train['abs_long'] =  train['abs_long'].astype(int)
train['abs_lat'] =  train['abs_lat'].astype(int)
train['pickup_longitude'] =  train['pickup_longitude'].astype(int)
###-----------Data Visualization--------------

sns.barplot(x='year',y='fare_amount',data=train)
plt.show()
#After closing the plot that opens in the last step,another plot opens automatically for the next step.
#This happens till the last plt.show() occurs.
sns.barplot(x='month',y='fare_amount',data=train)
plt.show()
sns.barplot(x='day',y='fare_amount',data=train)
plt.show()
sns.barplot(x='weekday',y='fare_amount',data=train)
plt.show()
sns.barplot(x='hour',y='fare_amount',data=train)
plt.show()

sns.barplot(x='month',y='passenger_count',data=train)
plt.show()
sns.barplot(x='day',y='passenger_count',data=train)
plt.show
sns.barplot(x='hour',y='passenger_count',data=train)
plt.show

print(test.shape)

###Detect outliers and dropping outliers
plt.boxplot(train['fare_amount'])
plt.show()
plt.boxplot(train['passenger_count'])
plt.show()
cnames=train.columns
def outliers_analysis(oa):
    for i in oa.columns:
        q75,q25 = np.percentile(oa.loc[:,i],[75,25])
        iqr=q75-q25
        min=q25-(iqr*1.5)
        max=q75+(iqr*1.5)
        oa = oa.drop(oa[oa.loc[:,i] < min].index)
        oa = oa.drop(oa[oa.loc[:,i] > max].index)
        return(oa)

train = outliers_analysis(train)
test = outliers_analysis(test)

print(test.shape)

#Splitting dataset keeping fare_amount variable aside
train_X = train.iloc[:,1:13] #all the independent variables
train_Y = train['fare_amount'] #only the dependent variable


##----------------------LINEAR REGRESSION-------------------------------------

import statsmodels.api as sm
LRmodel = sm.OLS(train_Y,train_X).fit()

print(LRmodel.summary())
#Check R-squared, adjusted R-squared etc.

predictions_LR = LRmodel.predict(test)

#MAPE
def MAPE(train_Y, predictions_LR):
    mapetest= np.mean(np.abs((train_Y - predictions_LR)/train_Y))*100
    return mapetest
print(MAPE(train_Y, predictions_LR))
    
#Since it gives infinite results we will use MAE and RMSE

#MAE
MAE = np.mean(abs(train_Y - predictions_LR))
print(MAE)
#3.2667889809326742


#RMSE
def RMSE(train_Y, predictions_LR):
    return np.sqrt(((predictions_LR - train_Y) ** 2).mean())
print(RMSE(train_Y, predictions_LR))
#4.1014448355035364

sns.barplot(x='hour',y='fare_amount',data=train)
plt.show()

##----------------------------Decision Tree--------------------------------

DTmodel = DecisionTreeRegressor(max_depth=2).fit(train_X, train_Y)
print(DTmodel)

predictions_DT = DTmodel.predict(test)
print(predictions_DT)

#Since it will give values in array let us convert the predictions_DT to dataframe in order to compute the MAE and RMSE
predictions_DT = pd.DataFrame(predictions_DT)
#MAE
MAEdt = np.mean(abs(train_Y - predictions_DT))
print(MAEdt)
#5.375968
#Accuracy :94.63

#RMSE
def RMSE(train_Y, predictions_DT):
    return np.sqrt(((predictions_DT - train_Y) ** 2).mean())
print(RMSE(train_Y, predictions_DT))

